zerver/lib/user_groups.py

```diff
@@ -119,6 +120,68 @@ def access_user_group_by_id(
     return user_group
 
 
+def access_user_group_for_deactivation(
+    user_group_id: int, user_profile: UserProfile
+) -> NamedUserGroup:
+    user_group = access_user_group_by_id(user_group_id, user_profile, for_read=False)
+
+    if (
+        user_group.direct_supergroups.exclude(named_user_group=None)
+        .filter(named_user_group__deactivated=False)
+        .exists()
+    ):
+        raise JsonableError(
+            _("You cannot deactivate a user group that is subgroup of any user group.")
+        )
+
+    anonymous_supergroup_ids = user_group.direct_supergroups.filter(
+        named_user_group=None
+    ).values_list("id", flat=True)
+
+    setting_group_ids_using_deactivating_user_group = [
+        *list(anonymous_supergroup_ids),
+        user_group.id,
+    ]
+
+    if Stream.objects.filter(
+        realm_id=user_group.realm_id,
+        deactivated=False,
+        can_remove_subscribers_group_id__in=setting_group_ids_using_deactivating_user_group,
+    ).exists():
+        raise JsonableError(_("You cannot deactivate a user group which is used for setting."))
+
+    if NamedUserGroup.objects.filter(
+        realm_id=user_group.realm_id,
+        deactivated=False,
+        can_mention_group_id__in=setting_group_ids_using_deactivating_user_group,
+    ).exists():
+        raise JsonableError(_("You cannot deactivate a user group which is used for setting."))
+
+    if (
+        Realm.objects.filter(id=user_group.realm_id)
+        .filter(
+            Q(create_multiuse_invite_group_id__in=setting_group_ids_using_deactivating_user_group)
+            | Q(can_access_all_users_group_id__in=setting_group_ids_using_deactivating_user_group)
+            | Q(
+                can_create_public_channel_group_id__in=setting_group_ids_using_deactivating_user_group
+            )
+            | Q(
+                can_create_private_channel_group_id__in=setting_group_ids_using_deactivating_user_group
+            )
+            | Q(
+                direct_message_initiator_group_id__in=setting_group_ids_using_deactivating_user_group
+            )
+            | Q(
+                direct_message_permission_group_id__in=setting_group_ids_using_deactivating_user_group
```

Comment:
Another approach that can be used here is that just fetch group ID for all the settings for the realm in a single query and then checking whether any IDs are common.

Reply:
I suspect that approach might produce cleaner SQL? Not sure, but `get_recursive_subgroups_for_groups` might prove helpful. But my main feedback on this code path is https://github.com/zulip/zulip/pull/30504/commits/447e81bcba3443ef635a9629645320a9165f6203#r1747772019

Reply:
Updated the code to loop over the settings, and that looks better.